feat: add C implementation for `stats/base/dists/geometric/mode`

0PrashantYadav0 opened the pull request:
Resolves #3638

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/geometric/mode along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3638 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C implementation for geometric mode

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/README.md:
```diff
@@ -120,6 +120,94 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_geometric_mode( p )
+
+Returns the mode of a geometric distribution.
```
```suggestion
Returns the [mode][mode] of a [geometric][geometric-distribution] distribution with success probability `p`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,68 @@
+	len = 100;
+	p = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		p[ i ] = ( randu() * 1.0 ) + EPS;
```
Can you please remove the `+ EPS` so the value is guaranteed to be in `[0,1]`?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/benchmark/benchmark.js:
```diff
@@ -31,14 +32,20 @@ var mode = require( './../lib' );
+	len = 100;
+	p = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		p[ i ] = ( randu() * 1.0 ) + EPS;
```
Can you please remove the `+ EPS` so the value is guaranteed to be in `[0,1]`?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,139 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		p[ i ] = random_uniform( 0.0, 1.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
Can you please remove the `+ STDLIB_CONSTANT_FLOAT64_EPS ` so the value is guaranteed to be in `[0,1]`?

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/src/main.c:
```diff
@@ -0,0 +1,41 @@
+		p < 0.0 ||
+		p > 1.0
+	) {
+		return 0.0 / 0.0;  // NaN
```
```suggestion
		return 0.0 / 0.0; // NaN
```

Planeshifter reviewed (changes requested):
Left a few comments concerning the generated `p` values; notice that you will have to update the `require`s / `include`s as well.

0PrashantYadav0 committed:
chore: completed the code review task

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
@Planeshifter Please review this pr

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/mode/lib/native.js:
```diff
@@ -0,0 +1,57 @@
+
+/**
+* Returns the mode of a geometric distribution.
+*
+* @param {Probability} p - success probability
```
```suggestion
*
* @private
* @param {Probability} p - success probability
```

Planeshifter committed:
docs: add missing private annotation

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (dismissed):
Looks good.

Planeshifter committed:
Merge branch 'develop' into geometric/mode

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
bench: require EPS

Planeshifter committed:
bench: use uniform instead of randu
